provider/boltmem/boltmem.go

```diff
@@ -284,25 +284,47 @@ func (s *Silences) Query(n, o int) (*types.SilencesQueryResponse, error) {
 		pageEnd = klen
 	}
 
-	silences := make([]*types.Silence, pageEnd-pageStart)
+	active, pending, elapsed := s.toBuckets()
+	sort.Sort(types.NewSilencesSorter(active, fn))
+	sort.Sort(types.NewSilencesSorter(pending, fn))
+	sort.Sort(types.NewSilencesSorter(elapsed, fn))
 
+	copiedSils := make([]*types.Silence, klen)
 	i := 0
-	for _, id := range s.k[pageStart:pageEnd] {
-		// We control the cache and the key so they shouldn't ever be
-		// out of sync, i.e. we don't need to worry about existence
-		// checks on the cache
-
-		// Make sure this silences[i] business is kosher and won't index out
-		// of bounds. Like with a test.
-		silences[i] = s.cache[id]
+	for _, sil := range active {
+		copiedSils[i] = sil
 		i++
 	}
+	for _, sil := range pending {
+		copiedSils[i] = sil
+		i++
+	}
+	for _, sil := range elapsed {
+		copiedSils[i] = sil
+		i++
+	}
+
 	return &types.SilencesQueryResponse{
-		Silences:      silences,
-		TotalSilences: klen,
+		Silences:      copiedSils[pageStart:pageEnd],
+		TotalSilences: len(copiedSils),
 	}, nil
 }
 
+func (s *Silences) toBuckets() (active []*types.Silence, pending []*types.Silence, elapsed []*types.Silence) {
```

Comment:
please suggest a better name if you want to stick with this method. happy with anything, maybe even just `s.buckets()`.